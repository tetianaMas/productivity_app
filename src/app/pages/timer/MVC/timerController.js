import { eventBus } from '../../../services/eventBus';

/**
 * @description controls timer component
 * @exports TimerController
 * @class TimerController
 */
export class TimerController {
  /**
   * @description Creates an instance of TimerController
   * @param {*} model instance of TimerModel
   * @param {*} view instance of TimerView
   * @memberof TimerController
   */
  constructor(model, view) {
    this.model = model;
    this.view = view;

    eventBus.subscribe('timer-loading', async () => {
      await this.model.init();
      eventBus.post('pageLoaded');
    });

    eventBus.subscribe('return-to-task-list', () => {
      const activeTask = this.model.activeTask;
      eventBus.post('end-task-working', activeTask);
    });

    eventBus.subscribe('raise-estimate', () => {
      this.model.raiseEstimate();
    });

    eventBus.subscribe('finish-pomodoro', isSuccessfullyCompleted => {
      this.model.finishPomodoro(isSuccessfullyCompleted);

      if (this.model.isTaskCompleted()) {
        eventBus.post('finish-task');
      }
    });

    eventBus.subscribe('finish-task', () => {
      this.model.finishTask();
    });
  }
}
