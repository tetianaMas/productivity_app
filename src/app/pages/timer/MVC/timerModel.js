import { Utils } from '../../../helpers/utils';
/**
 * @description manager timer component data
 * @exports TimerModel
 * @class TimerModel
 */
export class TimerModel {
  /**
   * @description Creates an instance of TimerModel
   * @param {object} dataService instance of DataService
   * @param {object} Observer instance of TimerObserver
   * @memberof TimerModel
   */
  constructor(dataService, Observer) {
    this.dataService = dataService;
    this.timerInitEvent = new Observer();
    this.initBreakEvent = new Observer();
    this.fillPomodoroEvent = new Observer();
    this.estimationChangedEvent = new Observer();
    this.taskCompletedEvent = new Observer();

    this.pomodoroSettings = null;
    this.activeTask = null;
    this.currentPomodoroIndex = 0;
    this.date = Utils.getCurrentDate();
  }

  /**
   * @description inits timer component
   * @memberof TimerModel
   */
  async init() {
    try {
      this.currentPomodoroIndex = 0;
      this.activeTask = this.getDataFromStorage('active-task');
      this.activeTask.failedPomodoros = [];
      this.activeTask.completedCount = [];

      const settings = await this.getAndSaveSettings();
      this.timerInitEvent.notify([this.activeTask, settings]);
    } catch (err) {
      console.error(err.message);
    }
  }

  /**
   * @description saves data from session storage
   * @param {string} key to get a task from session storage
   * @return {object} current task
   * @memberof TimerModel
   */
  getDataFromStorage(key) {
    if (!key || typeof key !== 'string') return;
    const task = this.dataService.getDataFromStorage(key);

    return task;
  }

  /**
   * @description gets and saves current settings
   * @return {object} settings
   * @memberof TimerModel
   */
  async getAndSaveSettings() {
    try {
      await this.dataService.receiveData('settings');
      this.pomodoroSettings = this.dataService.getDataFromStorage('settings');

      return this.pomodoroSettings;
    } catch (err) {
      console.error(err.message);
    }
  }

  /**
   * @description finishes pomodoro and shows notification
   * @param {boolean} isSuccessfullyCompleted true if successful
   * @memberof TimerModel
   */
  finishPomodoro(isSuccessfullyCompleted) {
    const isLongBreak = this.isLongBreak();

    isSuccessfullyCompleted
      ? this.setCompletedPomodoro()
      : this.setFailedPomodoro();

    this.initBreakEvent.notify([
      isSuccessfullyCompleted,
      isLongBreak,
      this.pomodoroSettings,
    ]);

    this.fillPomodoroEvent.notify([
      isSuccessfullyCompleted,
      this.currentPomodoroIndex,
    ]);
    this.raiseIndex();

    this.showFinishPomodoroNotification(false);
  }

  /**
   * @description adds failed pomodoros to current task
   * @memberof TimerModel
   */
  setFailedPomodoro() {
    this.activeTask.failedPomodoros.push(this.currentPomodoroIndex);
  }
  /**
   * @description adds successfully completed pomodoros to current task
   * @memberof TimerModel
   */
  setCompletedPomodoro() {
    this.activeTask.completedCount.push(this.currentPomodoroIndex);
  }

  /**
   * @description checks long break
   * @return {boolean} true if it's a long break
   * @memberof TimerModel
   */
  isLongBreak() {
    return (
      (this.currentPomodoroIndex || 1) %
        this.pomodoroSettings.iteration.value ===
      0
    );
  }

  /**
   * @description raises index of current task
   * @memberof TimerModel
   */
  raiseIndex() {
    this.currentPomodoroIndex++;
  }

  /**
   * @description raises estimation
   * @memberof TimerModel
   */
  raiseEstimate() {
    this.activeTask.estimation++;

    this.estimationChangedEvent.notify();
  }

  /**
   * @description checks task completion
   * @return {boolean} true if task completed
   * @memberof TimerModel
   */
  isTaskCompleted() {
    return (
      this.currentPomodoroIndex === parseInt(this.activeTask.estimation, 10)
    );
  }

  /**
   * @description finishes task, shows notification
   * @memberof TimerModel
   */
  finishTask() {
    this.activeTask.status = 'COMPLETED';
    this.activeTask.completeDate = this.date;
    if (!this.isTaskCompleted()) {
      this.completeAllPomodoros();
    }

    this.taskCompletedEvent.notify();
    this.dataService
      .sendData('tasks', this.activeTask, this.activeTask.id)
      .then(() => {
        this.showFinishPomodoroNotification(this.isLongBreak());
      })
      .catch(() => {
        this.showErrorNotification();
      });
  }

  /**
   * @description shows successful message
   * @param {boolean} isLongBreak
   * @memberof TimerModel
   */
  showFinishPomodoroNotification(isLongBreak) {
    $(document).notification('clean');

    if (isLongBreak) {
      $(document).notification({
        type: 'warning',
        text: 'Long break started, please have a rest!',
        showTime: 3,
      });
    } else {
      $(document).notification({
        type: 'success',
        text: 'You finished pomodoro!',
        showTime: 3,
      });
    }
  }

  /**
   * @description show error timer message
   * @memberof TimerModel
   */
  showErrorNotification() {
    $(document).notification('clean');
    $(document).notification({
      type: 'error',
      text: 'Unable to mark pomodoro/task as completed. Try again later',
      showTime: 3,
    });
  }

  /**
   * @description completes all pomodoros
   * @memberof TimerModel
   */
  completeAllPomodoros() {
    for (
      let i = this.currentPomodoroIndex;
      i < this.activeTask.estimation;
      i++
    ) {
      this.setCompletedPomodoro();
      this.fillPomodoroEvent.notify([true, i]);
      this.raiseIndex();
    }
  }
}
