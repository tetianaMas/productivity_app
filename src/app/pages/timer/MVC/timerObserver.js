/**
 * @description manages custom events for task-list component
 * @exports TimerObserver
 * @class TimerObserver
 */
export class TimerObserver {
  /**
   * @description Creates an instance of TimerObserver
   * @memberof TimerObserver
   */
  constructor() {
    this.observers = [];
  }

  /**
   * @description adds new callback to the observers
   * @param {object} handler callback to hangle event
   * @memberof TimerObserver
   */
  subscribe(handler) {
    this.observers.push(handler);
  }

  /**
   * @description emits all handlers
   * @param {object} data arguments for the handler
   * @memberof TimerObserver
   */
  notify(data) {
    this.observers.forEach(handler => handler(data));
  }
}
