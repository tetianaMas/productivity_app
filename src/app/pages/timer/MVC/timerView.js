import { eventBus } from '../../../services/eventBus';
import * as template from '../template/timer.hbs';

/**
 * @description manages timer view component
 * @exports TimerView
 * @class TimerView
 */
export class TimerView {
  /**
   * @description Creates an instance of TimerView
   * @param {*} model instance of TimerModel
   * @memberof TimerView
   */
  constructor(model) {
    this.model = model;
    this.MAX_AMOUNT_OF_POMODOROS = 10;

    this.model.timerInitEvent.subscribe(this.init.bind(this));
    this.model.initBreakEvent.subscribe(this.startBreakTimer.bind(this));
    this.model.fillPomodoroEvent.subscribe(
      this.renderCompletedPomodoro.bind(this)
    );
    this.model.estimationChangedEvent.subscribe(this.addPomodoro.bind(this));
    this.model.taskCompletedEvent.subscribe(
      this.renderCompletedScreen.bind(this)
    );
  }

  /**
   *
   * @description renders timer page, sets events
   * @param {object} arg task, settings
   * @memberof TimerView
   */
  init(arg) {
    const [task, settings] = arg;
    document.body.innerHTML = template(task);
    this.addEvents(settings);

    $('#js-main-timer').radialTimer({
      category: task.categoryId,
    });
  }

  /**
   * @description adds events
   * @param {object} settings current settings for pomodoro timer
   * @memberof TimerView
   */
  addEvents(settings) {
    document
      .querySelector('.js-btn-to-task-list')
      .addEventListener('click', () => eventBus.post('return-to-task-list'));

    document
      .querySelector('.js-start-timer')
      .addEventListener('click', () => this.startTimer(settings));

    document
      .querySelector('.js-btn-pomodoros-add')
      .addEventListener('click', () => eventBus.post('raise-estimate'));
  }

  /**
   * @description starts timer
   * @param {object} settings current settings for pomodoro timer
   * @memberof TimerView
   */
  startTimer(settings) {
    document.querySelector('.js-btn-to-task-list').classList.add('d-none');
    document.querySelector('.js-header-main').classList.add('hidden');

    this.renderTimerControlsButtons();
    this.startPomodoro(settings.work.value);
  }

  /**
   * @description renders timer controls buttons and sets events
   * @memberof TimerView
   */
  renderTimerControlsButtons() {
    document.querySelector('.js-main__control--timer').innerHTML = `
    <button class="main__control__button main__control__button--small 
    main__control__button--red js-fail-pomodora">
      fail pomodora
    </button>

    <button class="main__control__button main__control__button--small 
    main__control__button--mint js-finish-pomodora">
      finish pomodora
    </button>`;

    document
      .querySelector('.js-fail-pomodora')
      .addEventListener('click', () => this.finishPomodoro(false));

    document
      .querySelector('.js-finish-pomodora')
      .addEventListener('click', () => this.finishPomodoro(true));

    document.querySelector('.js-btn-pomodoros-add').classList.add('d-none');
  }

  /**
   * @description starts radial timer
   * @param {*} value
   * @memberof TimerView
   */
  startPomodoro(value) {
    $('#js-main-timer').radialTimer({
      time: value,
      content: timeLeft => {
        const currentTime = timeLeft - 1;

        if (currentTime > 0) {
          return `<p class="timer__time">${currentTime}</p>
          <p class="timer__dimension">min</p>`;
        }
      },
      renderInterval: 1,
      onTimeout: this.finishPomodoro.bind(TimerView),
    });
  }

  /**
   * @description finishes pomodoro
   * @param {boolean} state true if successfully finished
   * @memberof TimerView
   */
  finishPomodoro(state) {
    eventBus.post('finish-pomodoro', state);
  }

  /**
   * @description starts break timer
   * @param {object} arg
   * @memberof TimerView
   */
  startBreakTimer(arg) {
    const [isSuccessfullyCompleted, isLongBreak, settings] = arg;
    this.renderBreakControlsButtons(isSuccessfullyCompleted, settings);

    const options = {
      time: isLongBreak ? settings.longBreak.value : settings.shortBreak.value,
      content: timeLeft => {
        const currentTime = timeLeft - 1;

        if (currentTime > 0) {
          return `<p class="timer__break">break</p>
          <p class="timer__time">${currentTime}</p>
          <p class="timer__dimension">min</p>`;
        } else {
          return '<span class="timer__break--over">Break is over</span>';
        }
      },
      renderInterval: 1,
      onTimeout: function () {
        this.content();
      },
    };

    $('#js-main-timer').radialTimer(options);
  }

  /**
   * @description renders break timer buttons and sets events
   * @param {boolean} isSuccessfullyCompleted
   * @param {object} settings current settings for pomodoro timer
   * @memberof TimerView
   */
  renderBreakControlsButtons(isSuccessfullyCompleted, settings) {
    const buttonsBlock = document.querySelector('.js-main__control--timer');

    buttonsBlock.innerHTML = `
    <button class="main__control__button 
      main__control__button--mint main__control__button--small
      js-start-timer">
        start pomodora
    </button>`;

    document.querySelector('.js-btn-pomodoros-add').classList.remove('d-none');

    if (isSuccessfullyCompleted) {
      buttonsBlock.innerHTML += `
      <button class="main__control__button 
        main__control__button--small 
        main__control__button--blue js-finish-task">
          finish task
      </button>`;

      document
        .querySelector('.js-finish-task')
        .addEventListener('click', () => eventBus.post('finish-task'));
    }

    document
      .querySelector('.js-start-timer')
      .addEventListener('click', () => this.startTimer(settings));
  }

  /**
   * @description adds pomodoro
   * @memberof TimerView
   */
  addPomodoro() {
    const addPomodorosBtn = document.querySelector('.js-btn-pomodoros-add');
    const pomodorosBlock = document.querySelector('.js-timer-estimation-block');
    const pomodoro = this.createEmptyPomodoro();

    pomodorosBlock.insertBefore(pomodoro, addPomodorosBtn);

    if (pomodorosBlock.children.length > this.MAX_AMOUNT_OF_POMODOROS) {
      addPomodorosBtn.classList.add('hidden');
    }
  }

  /**
   * @description creates empty pomodoro
   * @return {object} empty pomodoro object
   * @memberof TimerView
   */
  createEmptyPomodoro() {
    const pomodoro = document.createElement('div');
    pomodoro.classList.add('main__task-info__icon');

    return pomodoro;
  }

  /**
   * @description renders completed pomodoro
   * @param {object} arg
   * @memberof TimerView
   */
  renderCompletedPomodoro(arg) {
    const [isSuccessfullyCompleted, index] = arg;
    const pomodorosBlock = document.querySelector('.js-timer-estimation-block');
    const currentPomodoro = pomodorosBlock.children[index];

    if (isSuccessfullyCompleted) {
      currentPomodoro.classList.add('main__task-info__icon--done');
    } else {
      currentPomodoro.classList.add('main__task-info__icon--failed');
    }
  }

  /**
   * @description renders full radial timer, manages main workflow buttons
   * @memberof TimerView
   */
  renderCompletedScreen() {
    $('#js-main-timer').radialTimer({
      content: 'You Completed Task',
      showFull: true,
    });

    document.querySelector('.js-btn-to-task-list').classList.remove('d-none');
    document.querySelector('.js-header-main').classList.remove('hidden');
    document.querySelector('.js-btn-to-reports').classList.remove('d-none');
    document.querySelector('.js-main__control--timer').innerHTML = '';
    document.querySelector('.js-btn-pomodoros-add').classList.add('d-none');
  }
}
