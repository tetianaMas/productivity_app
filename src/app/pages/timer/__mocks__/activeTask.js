export const activeTask = {
  categoryId: 'education',
  createDate: { day: 2, fullDeadline: 1617314006131, month: 'April' },
  deadline: 'April 3, 2021',
  deadlineDate: { day: '', fullDeadline: 1617397200000, month: 'today' },
  description: 'Test',
  failedPomodoros: [],
  completedCount: [],
  estimation: '3',
  id: '29118764318718169564',
  priority: 'urgent',
  status: 'ACTIVE',
  title: 'Test',
};
