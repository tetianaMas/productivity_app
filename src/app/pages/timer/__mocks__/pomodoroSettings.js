export const pomodoroSettings = {
  iteration: { default: 5, max: 5, min: 2, step: 1, value: 5 },
  longBreak: { default: 30, max: 30, min: 15, step: 5, value: 30 },
  shortBreak: { default: 5, max: 5, min: 3, step: 1, value: 5 },
  work: { default: 25, max: 25, min: 15, step: 5, value: 25 },
};
