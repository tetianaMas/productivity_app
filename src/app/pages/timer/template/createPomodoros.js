/**
 * @description gets amount of pomodoros and returns template for them
 * @exports
 * @param {string} amountOfPomodoros
 * @return {string} template for pomodoros
 */
export default function (amountOfPomodoros) {
  if (!amountOfPomodoros || !parseInt(amountOfPomodoros, 10)) return '';
  let fragment = '';

  for (let i = 0; i < parseInt(amountOfPomodoros, 10); i++) {
    fragment += '<div class="main__task-info__icon"></div>';
  }

  return fragment;
}
