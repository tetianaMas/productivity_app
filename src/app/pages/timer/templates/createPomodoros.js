/**
 * @description gets amount of pomodoros and returns template for them
 * @exports
 * @param {string} amountOfPomodoros
 * @return {string} template for pomodoros
 */
export default function (amountOfPomodoros) {
  let fragment = '';

  for (let i = 0; i < Number(amountOfPomodoros); i++) {
    fragment += '<div class="main__task-info__icon"></div>';
  }

  return fragment;
}
