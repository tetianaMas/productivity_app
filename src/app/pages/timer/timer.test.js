/* eslint-disable no-undef */
import { TimerModel } from './MVC/timerModel';
import { tasks } from '../tasks-list/__mocks__/tasks';
import timerTemplate from './template/timer.hbs';
import { TimerObserver } from './MVC/timerObserver';
import { dataService } from './../../services/dataService/dataService';
import createPomodoros from './template/createPomodoros';
import { pomodoroSettings } from './__mocks__/pomodoroSettings';
import { activeTask } from './__mocks__/activeTask';
import '../../components/notification/notification';

const timerModel = new TimerModel(dataService, TimerObserver);
timerModel.pomodoroSettings = pomodoroSettings;
timerModel.activeTask = activeTask;
describe('rendering', () => {
  test('renders correctly timer', () => {
    const tree = JSON.stringify(timerTemplate(tasks[0]));
    expect(tree).toMatchSnapshot();
  });
});

describe('createPomodoros', () => {
  test('should create pomodoros for template', () => {
    const res = createPomodoros('1');
    expect(res).toMatchSnapshot('<div class="main__task-info__icon"></div>');
  });

  test('should return falsy value', () => {
    const res = createPomodoros();
    expect(res).toBeFalsy();
  });

  test('should return falsy value', () => {
    const res = createPomodoros('');
    expect(res).toBeFalsy();
  });
});

describe('getDataFromStorage', () => {
  beforeEach(() => {
    spyOn(timerModel.dataService, 'getDataFromStorage');
  });
  test('should get data from storage', () => {
    timerModel.getDataFromStorage('active-task');
    expect(timerModel.dataService.getDataFromStorage).toHaveBeenCalledWith(
      'active-task'
    );
  });

  test('should return falsy value', () => {
    const res = timerModel.getDataFromStorage();
    expect(timerModel.dataService.getDataFromStorage).not.toBeCalled();
    expect(res).toBeFalsy();
  });

  test('should return falsy value', () => {
    const res = timerModel.getDataFromStorage(23425);
    expect(timerModel.dataService.getDataFromStorage).not.toBeCalled();
    expect(res).toBeFalsy();
  });
});

describe('finishPomodoro', () => {
  beforeAll(() => {
    spyOn(timerModel, 'isLongBreak');
    spyOn(timerModel, 'setCompletedPomodoro');
    spyOn(timerModel, 'setFailedPomodoro');
    spyOn(timerModel.initBreakEvent, 'notify');
    spyOn(timerModel.fillPomodoroEvent, 'notify');
    spyOn(timerModel, 'raiseIndex');

    timerModel.pomodoroSettings = pomodoroSettings;
  });
  test('finishes pomodoro and shows notification', () => {
    timerModel.finishPomodoro(true);

    expect(timerModel.isLongBreak).toBeCalled();
    expect(timerModel.setCompletedPomodoro).toBeCalled();
    expect(timerModel.initBreakEvent.notify).toBeCalled();
    expect(timerModel.fillPomodoroEvent.notify).toHaveBeenCalledWith([
      true,
      timerModel.currentPomodoroIndex,
    ]);
    expect(timerModel.raiseIndex).toBeCalled();
  });

  test('should return falsy result', () => {
    timerModel.finishPomodoro();
    expect(timerModel.setFailedPomodoro).toBeCalled();
  });
});

describe('setFailedPomodoro', () => {
  test('should set failed pomodoro', () => {
    spyOn(timerModel.activeTask.failedPomodoros, 'push');

    timerModel.setFailedPomodoro();
    expect(timerModel.activeTask.failedPomodoros.push).toHaveBeenCalledWith(
      timerModel.currentPomodoroIndex
    );
  });
});

describe('setCompletedPomodoro', () => {
  test('should set completed pomodoro', () => {
    spyOn(timerModel.activeTask.completedCount, 'push');

    timerModel.setCompletedPomodoro();
    expect(timerModel.activeTask.completedCount.push).toHaveBeenCalledWith(
      timerModel.currentPomodoroIndex
    );
  });
});

describe('isLongBreak', () => {
  test('should return true', () => {
    timerModel.currentPomodoroIndex = 1;
    timerModel.pomodoroSettings.iteration.value = 1;
    const res = timerModel.isLongBreak();

    expect(res).toBe(true);
  });

  test('should return false', () => {
    timerModel.currentPomodoroIndex = 3;
    timerModel.pomodoroSettings.iteration.value = 2;
    const res = timerModel.isLongBreak();

    expect(res).toBe(false);
  });
});

describe('raiseIndex', () => {
  test('should raise index', () => {
    let currentIndex = timerModel.currentPomodoroIndex;
    timerModel.raiseIndex();

    expect(timerModel.currentPomodoroIndex).toBe(currentIndex + 1);
  });
});

describe('raiseEstimate', () => {
  test('should raise estimate', () => {
    spyOn(timerModel.estimationChangedEvent, 'notify');

    let currentEstimation = parseInt(timerModel.activeTask.estimation, 10);
    timerModel.raiseEstimate();

    expect(timerModel.activeTask.estimation).toBe(currentEstimation + 1);
    expect(timerModel.estimationChangedEvent.notify).toBeCalled();
  });
});

describe('isTaskCompleted', () => {
  test('should return true', () => {
    timerModel.currentPomodoroIndex = 4;
    timerModel.activeTask.estimation = 4;
    const res = timerModel.isTaskCompleted();

    expect(res).toBe(true);
  });

  test('should return false', () => {
    timerModel.currentPomodoroIndex = 2;
    timerModel.activeTask.estimation = 4;
    const res = timerModel.isTaskCompleted();

    expect(res).toBe(false);
  });
});

describe('completeAllPomodoros', () => {
  beforeEach(() => {
    spyOn(timerModel, 'setCompletedPomodoro');
    spyOn(timerModel.fillPomodoroEvent, 'notify');
    spyOn(timerModel, 'raiseIndex');
  });
  test('should call func expected times', () => {
    timerModel.currentPomodoroIndex = 0;
    timerModel.activeTask.estimation = 2;
    timerModel.completeAllPomodoros();

    expect(timerModel.setCompletedPomodoro).toHaveBeenCalledTimes(2);
    expect(timerModel.fillPomodoroEvent.notify).toHaveBeenCalledTimes(2);
    expect(timerModel.raiseIndex).toHaveBeenCalledTimes(2);
  });

  test('should call func expected times', () => {
    timerModel.currentPomodoroIndex = 2;
    timerModel.activeTask.estimation = 2;
    timerModel.completeAllPomodoros();

    expect(timerModel.setCompletedPomodoro).not.toBeCalled();
    expect(timerModel.fillPomodoroEvent.notify).not.toBeCalled();
    expect(timerModel.raiseIndex).not.toBeCalled();
  });
});

describe('finishTask', () => {
  beforeAll(() => {
    spyOn(timerModel.taskCompletedEvent, 'notify');
    spyOn(timerModel, 'isTaskCompleted');
    timerModel.dataService.sendData = jest
      .fn()
      .mockImplementation(() => Promise.resolve());
    spyOn(timerModel, 'completeAllPomodoros');
    spyOn(timerModel, 'raiseIndex');
  });
  test('should finish task', () => {
    timerModel.currentPomodoroIndex = 2;
    timerModel.activeTask.estimation = '2';
    timerModel.isTaskCompleted = jest.fn().mockImplementation(() => true);

    timerModel.finishTask();
    expect(timerModel.activeTask.status).toBe('COMPLETED');
    expect(timerModel.activeTask.completeDate).toBe(timerModel.date);
    expect(timerModel.taskCompletedEvent.notify).toBeCalled();
    expect(timerModel.dataService.sendData).toHaveBeenCalledWith(
      'tasks',
      timerModel.activeTask,
      timerModel.activeTask.id
    );
    expect(timerModel.isTaskCompleted).toBeCalled();
    expect(timerModel.completeAllPomodoros).not.toBeCalled();
  });

  test('should call func expected times', () => {
    timerModel.currentPomodoroIndex = 0;
    timerModel.activeTask.estimation = '2';
    timerModel.isTaskCompleted = jest.fn().mockImplementation(() => false);

    timerModel.finishTask();

    expect(timerModel.activeTask.status).toBe('COMPLETED');
    expect(timerModel.activeTask.completeDate).toBe(timerModel.date);
    expect(timerModel.taskCompletedEvent.notify).toBeCalled();
    expect(timerModel.dataService.sendData).toHaveBeenCalledWith(
      'tasks',
      timerModel.activeTask,
      timerModel.activeTask.id
    );
    expect(timerModel.isTaskCompleted).toBeCalled();
    expect(timerModel.completeAllPomodoros).toBeCalled();
  });
});
