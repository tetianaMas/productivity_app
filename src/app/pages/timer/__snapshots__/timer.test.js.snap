// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createPomodoros should create pomodoros for template: <div class="main__task-info__icon"></div> 1`] = `"<div class=\\"main__task-info__icon\\"></div>"`;

exports[`rendering renders correctly timer 1`] = `"\\" <header class=\\\\\\"header js-header-main\\\\\\">\\\\r\\\\n\\\\r\\\\n <h1 class=\\\\\\"header__title\\\\\\">\\\\r\\\\n    daily task list\\\\r\\\\n  </h1>\\\\r\\\\n  <div class=\\\\\\"header__logo d-none\\\\\\">\\\\r\\\\n    <a href=\\\\\\"#\\\\\\">\\\\r\\\\n      <img src=\\\\\\"../images/Logo.svg\\\\\\" alt=\\\\\\"Logo\\\\\\" />\\\\r\\\\n    </a>\\\\r\\\\n  </div>\\\\r\\\\n  <nav class=\\\\\\"navigation\\\\\\" aria-label=\\\\\\"Main navigation\\\\\\">\\\\r\\\\n    <ul class=\\\\\\"navigation__list\\\\\\">\\\\r\\\\n      <li class=\\\\\\"navigation__item\\\\\\">\\\\r\\\\n        <button class=\\\\\\"navigation__btn js-header-nav-btn\\\\\\" data-link=\\\\\\"task-list\\\\\\" title=\\\\\\"Go to Tasks\\\\\\">\\\\r\\\\n          <span class=\\\\\\"icon-list \\\\\\"></span>\\\\r\\\\n        </button>\\\\r\\\\n      </li>\\\\r\\\\n      <li class=\\\\\\"navigation__item\\\\\\">\\\\r\\\\n        <button class=\\\\\\"navigation__btn js-header-nav-btn\\\\\\" data-link=\\\\\\"reports/day/tasks\\\\\\" title=\\\\\\"Go to Reports\\\\\\">\\\\r\\\\n          <span class=\\\\\\"icon-statistics \\\\\\"></span>\\\\r\\\\n        </button>\\\\r\\\\n      </li>\\\\r\\\\n      <li class=\\\\\\"navigation__item\\\\\\">\\\\r\\\\n        <button class=\\\\\\"navigation__btn js-header-nav-btn\\\\\\" data-link=\\\\\\"settings/pomodoros\\\\\\" title=\\\\\\"Go to Settings\\\\\\">\\\\r\\\\n          <span class=\\\\\\"icon-settings \\\\\\"></span>\\\\r\\\\n        </button>\\\\r\\\\n      </li>\\\\r\\\\n    </ul>\\\\r\\\\n  </nav>\\\\r\\\\n</header>\\\\r\\\\n<main class=\\\\\\"main main--static\\\\\\">\\\\r\\\\n    <div class=\\\\\\"main__task-info\\\\\\">\\\\r\\\\n      <h1 class=\\\\\\"main__task-info__title\\\\\\">Ывпыпа</h1>\\\\r\\\\n      <p class=\\\\\\"main__task-info__subtitle\\\\\\">Fgdgerf</p>\\\\r\\\\n\\\\r\\\\n      <div class=\\\\\\"main__task-info__estimation js-timer-estimation-block\\\\\\">\\\\r\\\\n        <div class=\\\\\\"main__task-info__icon\\\\\\"></div><div class=\\\\\\"main__task-info__icon\\\\\\"></div><div class=\\\\\\"main__task-info__icon\\\\\\"></div><div class=\\\\\\"main__task-info__icon\\\\\\"></div><div class=\\\\\\"main__task-info__icon\\\\\\"></div>\\\\r\\\\n        <button class=\\\\\\"icon-add d-none js-btn-pomodoros-add\\\\\\"></button>\\\\r\\\\n      </div>\\\\r\\\\n    </div>\\\\r\\\\n\\\\r\\\\n    <div class=\\\\\\"main__timer\\\\\\">\\\\r\\\\n      <a href=\\\\\\"/#/task-list\\\\\\" class=\\\\\\"main__timer__link main__timer__link--left icon-arrow-left js-btn-to-task-list\\\\\\" title=\\\\\\"Go to Global List\\\\\\">\\\\r\\\\n      </a>\\\\r\\\\n      <div id=\\\\\\"js-main-timer\\\\\\">\\\\r\\\\n    </div>\\\\r\\\\n      <a href=\\\\\\"/#/reports/day/tasks\\\\\\" class=\\\\\\"main__timer__link main__timer__link--right icon-arrow-right d-none js-btn-to-reports\\\\\\" title=\\\\\\"Go to Report\\\\\\">\\\\r\\\\n    </a>\\\\r\\\\n    </div>\\\\r\\\\n    \\\\r\\\\n    <div class=\\\\\\"main__control main__control--timer js-main__control--timer\\\\\\">\\\\r\\\\n      <button class=\\\\\\"main__control__button main__control__button--mint js-start-timer\\\\\\">\\\\r\\\\n        Start\\\\r\\\\n      </button>\\\\r\\\\n    </div>\\\\r\\\\n  </main>\\\\r\\\\n  \\""`;
